samples: matter: added thermostat sample application (#11737)
nrfconnect/sdk-nrf: samples/matter/thermostat/README.rst

.. _matter_thermostat_sample:

Matter: Thermostat
##################

.. contents::
   :local:
   :depth: 2

> This door lock sample demonstrates the usage of the :ref:`Matter <ug_matter>` application layer to build a thermostat device able to monitor temperature values.

Reviewer: ```suggestion
This thermostat sample demonstrates the usage of the :ref:`Matter <ug_matter>` application layer to build a thermostat device able to monitor and control temperature values.
```

> This basic implementation shows simple usage of Thermostat  enables the commissioning on the device, which allows it to join a Matter network built on top of a low-power, 802.15.4 Thread network or on top of a Wi-Fi network.

Reviewer: I'd suggest to remove this sentence or rephrase it.

Support for both Thread and Wi-Fi is mutually exclusive and depends on the hardware platform, so only one protocol can be supported for a specific Matter device.
In case of Thread, this device works as a Thread :ref:`Minimal End Device <thread_ot_device_types>`.

Use this sample as a reference for developing your own application.
See the :ref:`ug_matter_creating_accessory` page for an overview of the process you need to follow.

Requirements
************

The sample supports the following development kits:

.. table-from-sample-yaml::

For testing purposes, that is to commission the device and :ref:`control it remotely <matter_thermostat_network_mode>` through a Thread network, you also need a Matter controller device :ref:`configured on PC or smartphone <ug_matter_configuring>`. This requires additional hardware depending on the setup you choose.

.. note::
    |matter_gn_required_note|

IPv6 network support
====================

The development kits for this sample offer the following IPv6 network support for Matter:

* Matter over Thread is supported for ``nrf52840dk_nrf52840``, ``nrf5340dk_nrf5340_cpuapp``, and ``nrf21540dk_nrf52840``.

> * Matter over Wi-Fi is supported for ``nrf5340dk_nrf5340_cpuapp`` with the ``nrf7002_ek`` shield attached or for ``nrf7002dk_nrf5340_cpuapp``.

Reviewer 2: Is this sample supported for 7002?

Author: It is

As a result, the device can join the IPv6 network and communicate with other devices in the network.

.. _matter_thermostat_network_mode_onboarding:

Onboarding information
++++++++++++++++++++++

When you start the commissioning procedure, the controller must get the onboarding information from the Matter accessory device.
The onboarding information representation depends on your commissioner setup.

For this sample, you can use one of the following :ref:`onboarding information formats <ug_matter_network_topologies_commissioning_onboarding_formats>` to provide the commissioner with the data payload that includes the device discriminator and the setup PIN code:

  .. list-table:: Thermostat sample onboarding information
     :header-rows: 1

     * - QR Code
       - QR Code Payload
       - Manual pairing code
     * - Scan the following QR code with the app for your ecosystem:

         .. figure:: ../../../doc/nrf/images/matter_qr_code_thermostat.png
            :width: 200px
            :alt: QR code for commissioning the thermostat device

       - MT:O4CT3KQS02ZZ
       - 548G00

Configuration
*************

|config|

> 

Reviewer 2: These lines not needed.

Access Control List
===================

The Access Control List (ACL) is a list related to the Access Control cluster.
The list contains rules for managing and enforcing access control for a node's endpoints and their associated cluster instances.
In this sample's case, this allows the temperature measurement devices to receive messages from the thermostat and to get temperature data.

You can read more about ACLs on the :doc:`matter:access-control-guide` in the Matter documentation.

.. _matter_thermostat_sample_binding:

Binding

> =======
> .. include:: ../light_switch/README.rst

Reviewer 2: ```suggestion
=======

.. include:: ../light_switch/README.rst
```
Extra line needed.

   :start-after: matter_light_switch_sample_binding_start
   :end-before: matter_light_switch_sample_binding_end

> This sample can operate in two modes:

Reviewer: This piece is wrongly formatted:
https://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-11737/nrf/samples/matter/thermostat/README.html#binding
Please use bullet points.

>    mocked temperature sensor mode - the thermostat application generates the simulated temperature measurements
>    real temperature sensor mode - the thermostat is bound to the remote Matter temperature sensor, which provides real temperature measurements

Reviewer 2: ```suggestion
* Simulated temperature sensor mode - In this mode, the thermostat sample generates simulated temperature measurements.
* Real temperature sensor mode - In this mode, the thermostat sample is bound to the remote Matter temperature sensor, which provides real temperature measurements.
```

This sample supports the following build types, depending on the selected board:

* ``debug`` -- Debug version of the application - can be used to enable additional features for verifying the application behavior, such as logs or command-line shell.
* ``release`` -- Release version of the application - can be used to enable only the necessary application functionalities to optimize its performance.
* ``no_dfu`` -- Debug version of the application without Device Firmware Upgrade feature support - can be used for the nRF52840 DK, nRF5340 DK, nRF7002 DK, and nRF21540 DK.

.. note::
    `Selecting a build type`_ is optional.
    The ``debug`` build type is used by default if no build type is explicitly selected.

Device Firmware Upgrade support
===============================

.. include:: ../lock/README.rst
    :start-after: matter_door_lock_sample_build_with_dfu_start
    :end-before: matter_door_lock_sample_build_with_dfu_end

FEM support
===========

.. include:: /includes/sample_fem_support.txt

User interface
**************

.. include:: ../lock/README.rst
    :start-after: matter_door_lock_sample_led1_start
    :end-before: matter_door_lock_sample_led1_end

> Button 1:

Reviewer: Please follow the lock's user interface description:
https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/samples/matter/lock/README.html#user-interface
i.e. use bullet points to clearly separate buttons functions per DK (distinguish between nRF52/53 and nRF7002 DKs) and different interactions with buttons (pressing for a given time period).

-------------------------------

.. include:: /config_and_build/modifying.rst
   :start-after: build_types_selection_vsc_start
   :end-before: build_types_selection_vsc_end

Selecting a build type from command line
----------------------------------------

.. include:: /config_and_build/modifying.rst
   :start-after: build_types_selection_cmd_start
   :end-before: For example, you can replace the

For example, you can replace the *selected_build_type* variable to build the ``release`` firmware for ``nrf52840dk_nrf52840`` by running the following command in the project directory:

.. parsed-literal::
   :class: highlight

   west build -b nrf52840dk_nrf52840 -d build_nrf52840dk_nrf52840 -- -DCONF_FILE=prj_release.conf

The ``build_nrf52840dk_nrf52840`` parameter specifies the output directory for the build files.

.. note::
   If the selected board does not support the selected build type, the build is interrupted.
   For example, if the ``shell`` build type is not supported by the selected board, the following notification appears:

   .. code-block:: console

      File not found: ./ncs/nrf/samples/matter/thermostat/configuration/nrf52840dk_nrf52840/prj_shell.conf

> 

Reviewer 2: Lines not needed.

Testing
=======

When you have built the sample and programmed it to your development kit, it automatically starts the Bluetooth LE advertising and the **LED1** starts flashing (Short Flash On).

> Thermostat should update its values every 30s and print data to logs.

Reviewer: This should be rephrased, maybe something like:
```suggestion
Thermostat will print measured temperature value to the UART console with 30 seconds interval.
```

By pressing **Button 2** you can print thermostat data to the logs.
At this point, you can press **Button 1** for six seconds to initiate the factory reset of the device.

.. _matter_thermostat_network_testing:

Testing in a network
--------------------

To test the sample in a Matter-enabled Thread network, complete the following steps:

.. matter_thermostat_sample_testing_start

1. |connect_kit|
#. |connect_terminal_ANSI|
#. Press **Button 2** to print Thermostat data to logs.
#. Press **Button 1** for six seconds to initiate the factory reset of the device.

The device reboots after all its settings are erased.

Testing with external sensor
----------------------------

>    Thermostat sample can be connected with external sensor device (for example Thingy:53: Matter weather station). Connection is done by :ref:`matter_thermostat_sample_binding` to matter temperature measurement cluster.

Reviewer: ```suggestion
   Thermostat sample can be connected with external sensor device (for example Thingy:53: Matter weather station). Connection is done by :ref:`matter_thermostat_sample_binding` to Matter temperature measurement cluster.
```
Maybe we should link the:
[Thingy:53: Matter weather station](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/applications/matter_weather_station/README.html)

Also the "_Connection is done by"_ sounds a bit weird, but maybe @greg-fer will comment on that.

>    Thermostat sample can be connected with external sensor device (for example Thingy:53: Matter weather station). Connection is done by :ref:`matter_thermostat_sample_binding` to matter temperature measurement cluster.

Author: Im not sure how to do that

>    Thermostat sample can be connected with external sensor device (for example Thingy:53: Matter weather station). Connection is done by :ref:`matter_thermostat_sample_binding` to matter temperature measurement cluster.

Reviewer 3: I guess @greg-fer may help here.

   .. note::
      To enable external sensor usage, you need to set CONFIG_EXTERNAL_SENSOR=y in .conf file
      or add -DCONFIG_EXTERNAL_SENSOR=y flag to build west command

1. |connect_kit|
#. |connect_terminal_ANSI|
#. Commission both devices into a Matter network by following the guides linked on the :ref:`ug_matter_configuring` page for the Matter controller you want to use.
   The guides walk you through the following steps:

   * Only if you are configuring Matter over Thread: Configure the Thread Border Router.
   * Build and install the Matter controller.
   * Commission the device.
     You can use the :ref:`matter_thermostat_network_mode_onboarding` listed earlier on this page.

> #. Connect_thermostat_with_sensor

Reviewer: Why using `_`s? Anyway, I guess we need to clarify this point.

#. Use the :doc:`CHIP Tool <matter:chip_tool_guide>` ("Writing ACL to the ``accesscontrol`` cluster" section) to add proper ACL for the temperature device.

>    * chip-tool accesscontrol write acl '[{"fabricIndex": 1, "privilege": 5, "authMode": 2, "subjects": [112233], "targets": null}, \
>      {"fabricIndex": 1, "privilege": 1, "authMode": 2, "subjects": [<Thermostat NodeId>], "targets": \
>      [{"cluster": 1026, "endpoint": <TemperatureSensor EndpointId>, "deviceType": null}]}]' <Sensor Device NodeId> 0

Reviewer 2: This needs to be formatted as a ``.. code-block::``:
```suggestion
   .. code-block::
   
   chip-tool accesscontrol write acl '[{"fabricIndex": 1, "privilege": 5, "authMode": 2, "subjects": [112233], "targets": null}, \
     {"fabricIndex": 1, "privilege": 1, "authMode": 2, "subjects": [<Thermostat NodeId>], "targets": \
     [{"cluster": 1026, "endpoint": <TemperatureSensor EndpointId>, "deviceType": null}]}]' <Sensor Device NodeId> 0
```

#. Use the :doc:`CHIP Tool <matter:chip_tool_guide>` ("Adding a binding table to the ``binding`` cluster) to add binding table to thermostat.

   * chip-tool binding write binding '[{"fabricIndex": 1, "node": <TemperatureSensor NodeId>, "endpoint": <TemperatureSensor EndpointId>, \

>      "cluster": 1026}]' <Thermostat NodeId> <Thermostat EndpointId>

Reviewer 2: Same as above: command needs to go into a code-block.

   After that thermostat should be able to read data from temperature sensor.

> 

Reviewer 2: 1 line not needed here.

> #. Press **Button 2** to print Thermostat data to logs.

Reviewer: ```suggestion
#. Press **Button 2** to print Thermostat data to the UART console.
```

      $ ./chip-tool accesscontrol write acl '[{"fabricIndex": 1, "privilege": 5, "authMode": 2, "subjects": [112233], "targets": null}, \
      {"fabricIndex": 1, "privilege": 1, "authMode": 2, "subjects": [<thermostat_node_ID>], "targets": \
      [{"cluster": 1026, "endpoint": <thermostat_sensor_endpoint_ID>, "deviceType": null}]}]' <Sensor Device NodeId> 0

#. Write a binding table to the thermostat to inform the device about the temperature sensor endpoint.
   Use the following command, with *<temperature_sensor_node_ID>* and *<thermostat_sensor_endpoint_ID>* values from the previous step about commissioning:

   .. code-block:: console

      $ ./chip-tool binding write binding '[{"fabricIndex": 1, "node": <temperature_sensor_node_ID>, "endpoint": <thermostat_sensor_endpoint_ID>, \
      "cluster": 1026}]' <thermostat_node_ID> 1

   (You can read more about this step in the "Adding a binding table to the ``binding`` cluster" in the CHIP Tool guide.)

   The thermostat is now able to read the real temperature data from the temperature sensor device.
   The connection is ensured by :ref:`matter_thermostat_sample_binding` to Matter's temperature measurement cluster.

> #. Press **Button 2** to print the temperature data from the temperature sensor device to the UART terminal.

Reviewer: This sentence may suggest that we print the values directly from the temperature sensor device which is not true.
```suggestion
#. Press **Button 2** to print the most recent temperature data from the thermostat to the UART terminal.
```

   :local:
   :depth: 2

This thermostat sample demonstrates the usage of the :ref:`Matter <ug_matter>` application layer to build a thermostat device for monitoring temperature values.
This device works as a Matter accessory device, meaning it can be paired and controlled remotely over a Matter network built on top of a low-power, 802.15.4 Thread network or on top of a Wi-Fi network.
In case of Thread, this device works as a Thread :ref:`Minimal End Device <thread_ot_device_types>`.
Support for both Thread and Wi-Fi is mutually exclusive and depends on the hardware platform, so only one protocol can be supported for a specific Matter device.

Additionally, this example allows you to connect to a temperature sensor device that can also be used for temperature measurement.

Requirements
************

The sample supports the following development kits:

.. table-from-sample-yaml::

If you want to commission the device and :ref:`control it remotely <matter_thermostat_network_mode>` in a Thread network, you also need a Matter controller device :ref:`configured on PC or smartphone <ug_matter_configuring>`.
This requires additional hardware depending on the setup you choose.

> Similarly, if you want to test the sample with :ref:`matter_thermostat_sample_sensor`, you need additional hardware that integrates a temperature sensor.

Reviewer 2: Can we just use "incorporates"? Three verbs is a bit too much.

> Similarly, if you want to test the sample with :ref:`matter_thermostat_sample_sensor`, you need additional hardware that integrates a temperature sensor.

Reviewer: Sure, I just wanted to provide some options here.

  This mode requires :ref:`matter_thermostat_sample_sensor`.

The sample uses buttons for printing temperature measurement results to the terminal.

You can test the sample in the following ways:

* Standalone, using a single DK that runs the thermostat application.
* Remotely over the Thread or the Wi-Fi protocol, which in either case requires more devices, including a Matter controller that you can configure either on a PC or a mobile device.

You can enable both methods after :ref:`building and running the sample <matter_thermostat_sample_remote_control>`.

Configuration
*************

|config|

Access Control List
===================

The Access Control List (ACL) is a list related to the Access Control cluster.
The list contains rules for managing and enforcing access control for a node's endpoints and their associated cluster instances.
In this sample's case, this allows the temperature measurement devices to receive messages from the thermostat and to get temperature data.

You can read more about ACLs on the :doc:`matter:access-control-guide` in the Matter documentation.

.. _matter_thermostat_sample_sensor:

External sensor integration
===========================

> The thermsotat sample lets you connect to an external temperature sensor, for example :ref:`Matter weather station application on Nordic Thingy:53 <matter_weather_station_app>`.

Reviewer 2: What do you mean by the actual codebase?

> The thermsotat sample lets you connect to an external temperature sensor, for example :ref:`Matter weather station application on Nordic Thingy:53 <matter_weather_station_app>`.

Reviewer 2: That is ok, it will work fine on live.

Device Firmware Upgrade support
===============================

.. include:: ../lock/README.rst
    :start-after: matter_door_lock_sample_build_with_dfu_start
    :end-before: matter_door_lock_sample_build_with_dfu_end

.. _matter_thermostat_network_mode:

Remote testing in a network
===========================

.. include:: ../light_bulb/README.rst
    :start-after: matter_light_bulb_sample_remote_testing_start
    :end-before: matter_light_bulb_sample_remote_testing_end

User interface
**************

.. include:: ../lock/README.rst
    :start-after: matter_door_lock_sample_led1_start
    :end-before: matter_door_lock_sample_led1_end

Button 1:
    Depending on how long you press the button:

    * If pressed for less than three seconds, it initiates the SMP server (Security Manager Protocol).
      After that, the Direct Firmware Update (DFU) over Bluetooth Low Energy can be started.
      (See `Upgrading the device firmware`_.)
    * If pressed for more than three seconds, it initiates the factory reset of the device.
      Releasing the button within the 3-second window cancels the factory reset procedure.

Button 2:

>     * On nRF52840 DK, nRF5340 DK, and nRF21540 DK: Prints the thermostat data to terminal.

Reviewer: Again, I guess we don't support nRF21540 DK, or do we?

.. _matter_thermostat_sample_basic_features_tests:

Testing basic features
----------------------

After building the sample and programming it to your development kit, complete the following steps to test its basic features:

1. |connect_kit|
#. |connect_terminal_ANSI|
#. Observe that the **LED1** starts flashing (short flash on).
   This means that the sample has automatically started the Bluetooth LE advertising.
#. Observe the UART terminal.
   The sample starts automatically printing the simulated temperature data to the terminal with 30-second intervals.
#. Press **Button 2** to print the thermostat temperature data to the terminal.
#. Press **Button 1** for six seconds to initiate the factory reset of the device.

The device reboots after all its settings are erased.

.. _matter_thermostat_sensor_testing:

Testing with external sensor
----------------------------

After building this sample and the :ref:`Matter weather station <matter_weather_station_app>` application and programming them to the development kits, complete the following steps to test communication between both devices:

> 1. |connect_kit|

Reviewer: "Connect the kit to the computer using a USB cable" -> which kit? I guess we should clarify which kit is which (DK vs. Thingy).

Reviewer 2: Thingy is a "prototyping platform", so I guess kit = DK. This means line 248 needs also an edit.

#. |connect_terminal_ANSI|
#. If devices were not erased during the programming, press the button responsible for the factory reset on each device.
#. On each device, press the button that starts the Bluetooth LE advertising.
#. Commission devices to the Matter network.
   See `Commissioning the device`_ for more information.
   During the commissioning process, write down the values for the thermostat node ID, the temperature sensor node ID, and the temperature sensor endpoint ID.
   These IDs are going to be used in the next steps (*<thermostat_node_ID>*, *<temperature_sensor_node_ID>*, and *<thermostat_sensor_endpoint_ID>*, respectively).
#. Use the :doc:`CHIP Tool <matter:chip_tool_guide>` ("Writing ACL to the ``accesscontrol`` cluster" section) to add proper ACL for the temperature sensor device.

>    Use the following command, with *<thermostat_node_ID>* and *<temperature_sensor_endpoint_ID>* values from the previous step about commissioning:

Reviewer: Maybe these template names should directly reflect the ones used in the command template? I mean e.g. `<thermostat_node_ID>` vs. `<Thermostat NodeId>`

Reviewer 2: Can we edit the command template? Every other piece of Matter docs uses underscored variables.

   .. code-block:: console

      $ ./chip-tool accesscontrol write acl '[{"fabricIndex": 1, "privilege": 5, "authMode": 2, "subjects": [112233], "targets": null}, \
      {"fabricIndex": 1, "privilege": 1, "authMode": 2, "subjects": [<Thermostat NodeId>], "targets": \
      [{"cluster": 1026, "endpoint": <TemperatureSensor EndpointId>, "deviceType": null}]}]' <Sensor Device NodeId> 0

#. Write a binding table to the thermostat to inform the device about the temperature sensor endpoint.

>    Use the following command, with *<temperature_sensor_node_ID>* and *<thermostat_sensor_endpoint_ID>* values from the previous step about commissioning:

Reviewer: Same as above.

   .. code-block:: console

      $ ./chip-tool binding write binding '[{"fabricIndex": 1, "node": <TemperatureSensor NodeId>, "endpoint": <TemperatureSensor EndpointId>, \
      "cluster": 1026}]' <Thermostat NodeId> 1

   (You can read more about this step in the "Adding a binding table to the ``binding`` cluster" in the CHIP Tool guide.)

   The thermostat is now able to read the real temperature data from the temperature sensor device.
   The connection is ensured by :ref:`matter_thermostat_sample_binding` to Matter's temperature measurement cluster.

> #. Press **Button 2** to print the temperature data from the temperature sensor device to the UART terminal.

Reviewer: I guess it would be good to explicitly explain somewhere if the act of pressing the button initiates the read-out of data from remote device and immediately forwards the result, or it just returns the recent measured value stored in the thermostat (with the temperature sensor providing the data periodically).

> #. Press **Button 1** for six seconds to initiate the factory reset of the thermostat device.

Reviewer: Not sure if this is relevant to "testing with external sensor", it more fits to the smoke testing already described in "Testing basic features" section.

Reviewer 2: Removed.
